psi4/driver/p4util/fcidump.py

```diff
@@ -211,26 +211,170 @@ def write_eigenvalues(eigs, mo_idx):
+    eri[(idxs[sl, 2], idxs[sl, 3], idxs[sl, 0], idxs[sl, 1])] = ints[sl]
+    eri[(idxs[sl, 3], idxs[sl, 2], idxs[sl, 0], idxs[sl, 1])] = ints[sl]
+    eri[(idxs[sl, 2], idxs[sl, 3], idxs[sl, 1], idxs[sl, 0])] = ints[sl]
+    eri[(idxs[sl, 3], idxs[sl, 2], idxs[sl, 1], idxs[sl, 0])] = ints[sl]
+    intdump['eri'] = eri
+
+    return intdump
+
+
+def compare_fcidump_headers(expected, computed, label):
+    """Function to compare the headers of two FCIDUMP files. Prints :py:func:`util.success`
     when value *computed* matches value *expected*.
     Performs a system exit on failure. Used in input files in the test suite.
+
+    :returns: a dictionary of energies computed from the MO integrals.
+    The key-value pairs are:
+      - 'NUCLEAR REPULSION ENERGY' : nuclear repulsion plus frozen core energy
+      - 'ONE-ELECTRON ENERGY' : SCF one-electron energy
+      - 'TWO-ELECTRON ENERGY' : SCF two-electron energy
+      - 'SCF TOTAL ENERGY' : SCF total energy
+      - 'MP2 CORRELATION ENERGY' : MP2 correlation energy
+
+    :param expected: reference FCIDUMP file
+    :param computed: computed FCIDUMP file
+    :param label: string labelling the test
     """
     # Grab expected header and integrals
-    with open(expected) as ref_dump:
-        ref_header = [next(ref_dump) for x in range(7)]
-    ref_intdump = np.loadtxt(expected, skiprows=7)
-    # Grab computed header and integrals
-    with open(computed) as dump:
-        header = [next(dump) for x in range(7)]
-    intdump = np.loadtxt(computed, skiprows=7)
+    ref_intdump = readin_fcidump(expected)
+    intdump = readin_fcidump(computed)
 
     # Compare headers
-    header_diff = list(unified_diff(ref_header, header, fromfile=expected, tofile=computed))
+    header_diff = DeepDiff(
+        ref_intdump,
+        intdump,
+        ignore_order=True,
+        exclude_paths={"root['enuc']", "root['hcore']", "root['eri']", "root['epsilon']"})
     if header_diff:
         message = ("\tComputed FCIDUMP file header does not match expected header.\n")
-        for l in header_diff:
-            message += l
-        raise TestComparisonError(message)
-    # Compare integrals
-    compare_arrays(ref_intdump, intdump, 10, label)
+        raise TestComparisonError(header_diff)
+    success(label)
+
+    energies = {}
+    energies['NUCLEAR REPULSION ENERGY'] = intdump['enuc']
+    epsilon = intdump['epsilon']
+    Hcore = intdump['hcore']
+    eri = intdump['eri']
+
+    # Compute SCF energy
+    energies['ONE-ELECTRON ENERGY'], energies['TWO-ELECTRON ENERGY'] = _scf_energy_from_fcidump(
+        Hcore, eri,
+        np.where(epsilon < 0)[0], intdump['uhf'])
+    energies[
+        'SCF TOTAL ENERGY'] = energies['ONE-ELECTRON ENERGY'] + energies['TWO-ELECTRON ENERGY'] + energies['NUCLEAR REPULSION ENERGY']
+
+    # Compute MP2 energy
+    energies['MP2 CORRELATION ENERGY'] = _mp2_energy_from_fcidump(eri, epsilon, intdump['uhf'])
+
```

Comment:
why not compute the energy quantities from both ref and computed fcidumps, then `compare_values` the pieces?

Reply:
My thinking was that comparing with the energies computed by Psi4 itself would be stringent enough. I could add the recomputed-from-reference-FCIDUMP _vs_ recomputed-from-fresh-FCIDUMP test internally in the `compare_fcidumps` function. I just don't think it would be more informative.

Reply:
It wouldn't be any more stringent than what you have now. I just figured it'd be more self-contained and thorough. call `compare_values` on all quantities within a single `compare_fcidumps` and then in the test case a single `compare_fcidumps` suffices for each segment and maybe just spot-check a single energy from the returned energies dict. Could also help maintain the `compare_*` fns pattern of returning None (pass) or raise Exception (fail). For pytest testing, I have the fns wrapped to convert to True/False.

Up to you.

Reply:
I've updated following your suggestion. I agree we need to keep consistency with the other `compare_*` functions.